designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala

```diff
@@ -0,0 +1,129 @@
+  }
+
+  private def checkForMissingParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    val paramsFromRequest = requestParamsMap.keys.toSet
+    val mandatoryParams = customActionParams.collect {
+      case param if param.validators.contains(MandatoryParameterValidator) && !paramsFromRequest.contains(param.name) =>
+        MandatoryParameterValidator.isValid(ParameterName(param.name), Expression.spel(""), None, None)
+    }
+    mandatoryParams match {
+      case Nil      => Validated.Valid(())
+      case nonEmpty => nonEmpty.traverse(_.toValidatedNel).map(_ => ())
+    }
+  }
+
+  private def validateParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    requestParamsMap match {
+      case paramsMap if paramsMap.nonEmpty =>
+        handleNonEmptyParamsRequest(paramsMap, customActionParams)
+      case emptyMap if emptyMap.isEmpty =>
+        handleEmptyParamsRequest(customActionParams)
+    }
+  }
+
+  private def handleNonEmptyParamsRequest(
+      paramsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    paramsMap.toList.map { case (name, expression) =>
+      customActionParams.find(_.name == name) match {
+        case Some(param) => toValidatedNel(param, expression, name)
+        case None =>
+          MismatchParameter(
+            s"Couldn't find a matching parameter in action definition for this param: $name",
+            "",
+            ParameterName(name),
+            nodeId.id
+          ).invalidNel[Unit]
+      }
+    }.sequence_
+  }
+
+  private def toValidatedNel(param: CustomActionParameter, expressionValue: String, parameterName: String)(
+      implicit nodeId: NodeId
+  ): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+    param.validators
+      .map { validator =>
+        validator.isValid(
+          paramName = ParameterName(parameterName),
+          expression = Expression.spel(expressionValue),
+          value = Some(expressionValue),
+          label = None
+        )
+      }
+      .traverse(_.toValidatedNel)
+      .map(_ => ())
+  }
+
+  private def handleEmptyParamsRequest(
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    customActionParams
+      .collect { case param if param.validators.nonEmpty => (param.name, param.validators) }
+      .flatMap { case (name, validators) =>
+        validators.collect { case validator => validator }.map {
+          _.isValid(
+            paramName = ParameterName(name),
+            expression = Expression.spel("None"),
```

Comment:
What the heck? Why do you put some `None` into spel expression?

Reply:
It's an attempt to validate an empty expression when the user doesn't input anything. Is there some more subtle way?

Reply:
I don't understand your answer. Why do you validate an empty expression when the user doesn't input anything?